File: frecuencyCounterPattern/anagram/anagram.py
def valid_anagram(first_str, secon_str):
    if len(first_str) != len(secon_str):
        return False

    lookup = {}
    for char in first_str:
        lookup[char] = lookup.get(char, 0) + 1

    for char in secon_str:
        if not lookup.get(char):
            return False
        lookup[char] -= 1
    return True


print(valid_anagram('', '')) # True
print(valid_anagram('azz', 'zaa')) # False
print(valid_anagram('anagram', 'nagaram')) # True
print(valid_anagram('rat', 'car')) # False
print(valid_anagram('awesome', 'awesom')) # False
print(valid_anagram('qwerty', 'qeywrt'))# True
print(valid_anagram('texttwisttime', 'timetwisttext')) # True

File: frecuencyCounterPattern/anagram/anagram.rb
def valid_anagram(first_str, second_str)
    return false if first_str.length != second_str.length

    lookup = {}
    first_str.each_char {|char| lookup[char] = (lookup[char] || 0) + 1}

    second_str.each_char do |char|
        return false if !lookup.include?(char)
        return false if lookup[char].zero?
        lookup[char] -= 1
    end
    return true
end

puts(valid_anagram('', '')) # true
puts(valid_anagram('azz', 'zaa')) # false
puts(valid_anagram('anagram', 'nagaram')) # true
puts(valid_anagram('rat', 'car')) # false
puts(valid_anagram('awesome', 'awesom')) # false
puts(valid_anagram('qwerty', 'qeywrt'))# true
puts(valid_anagram('texttwisttime', 'timetwisttext')) # true